src/ParserCombinators/Parser.hs

```diff
@@ -0,0 +1,122 @@
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE InstanceSigs #-}
+
+module ParserCombinators.Parser
+  ( token
+  , satisfy
+  , eof
+  , char
+  , string
+  , between
+  , space
+  , spaces
+  , spaces1
+  , bool
+  , digit
+  , integer
+  , symbol
+  , many1
+  , try
+  , sepBy
+  , sepBy1
+  ) where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (void)
+import Data.Char (isAlphaNum, isDigit)
+import Data.Foldable (asum)
+
+import ParserCombinators.Datatypes
+
+-- | Parse a token/character using a custom error generator and condition
+token :: (i -> ParseErrorType i e) -> (i -> Bool) -> Parser i [] e i
+token mkErr predicate = Parser $ \input offset ->
+  case input of
+    [] -> Left $ pure $ ParseError offset EndOfInput
+    x : rest
+      | predicate x -> Right (succ offset, x, rest)
+      | otherwise -> Left $ pure $ ParseError offset $ mkErr x
+
+-- | Parse taken/character by a condition
+satisfy :: (i -> Bool) -> Parser i [] e i
+satisfy = token Unexpected
+
+-- | Parse EOF
+eof :: Parser i [] e ()
+eof = Parser $ \input offset ->
+  case input of
+    [] -> Right (offset, (), [])
+    x : _ -> Left [ParseError offset $ ExpectedEndOfFile x]
```

Comment:
You don't need to use the `Parser` constructor to implement this, I think.

Reply:
I think I do, using `token` assumes that the parser must fail on empty input which doesn't apply to `eof`